### PR TITLE
Allow administrator to grant administration rights to a user

### PR DESCRIPTION
**Pull request by romain-ortega**
An admin can't change his own rank.

### CODE REVIEW

File: nanocloud/routes/users/users.go [hunk 1]
```diff
@@ -128,18 +128,28 @@ func Disable(userId string) (int, error) {
...
 
 	user, err := users.GetUser(u.GetID())
 	if err != nil {
 		return apiErrors.UserNotFound
 	}
 
-	if u.Password != "" {
+	if u.IsAdmin != user.IsAdmin {
+		if u.Id == user.Id {
+			return apiErrors.Unauthorized.Detail("You can't change your own rank")
```
Comment (dynamiccast) [inline nanocloud/routes/users/users.go]: Please change can't -> cannot

File: nanocloud/routes/users/users.go [hunk 2]
```diff
@@ -128,18 +128,28 @@ func Disable(userId string) (int, error) {
...
+		}
+		err = users.UpdateUserRank(user.GetID(), u.IsAdmin)
+		if err != nil {
+			log.Error(err)
+			return apiErrors.InternalError.Detail("Unable to update the rank.")
```
Comment (dynamiccast) [inline nanocloud/routes/users/users.go]: No dot at the end of the sentence.

File: webapp/app/protected/users/user/controller.js [hunk 3]
```diff
@@ -16,101 +19,128 @@ export default Ember.Controller.extend({
...
+          this.model.save()
+            .then(() => {
+              this.send('refreshModel');
+              this.toast.success('Rank has been updated successfully');
+            }, () => {
+              this.toast.error("Rank hasn't been updated");
+            }
+            );
+        }, (err) => {
+          console.error(err);
```
Comment (dynamiccast) [inline webapp/app/protected/users/user/controller.js]: Catch the error and display the error with a toast

File: nanocloud/models/users/users.go [hunk 4]
```diff
@@ -105,9 +107,9 @@ func FindUsers() ([]*User, error) {
 			&user.Email,
 			&user.IsAdmin,
 			&user.Activated,
-			&user.SignupDate,
+			&timestamp,
 		)
-		user.SignupDate *= 1000
+		user.SignupDate = int(1000 * timestamp)
```
Comment (Gentux) [inline nanocloud/models/users/users.go:113]: Magic value here.
I suggest you to put a comment on this line to indicate why you multiply by a thousand your signupDate

File: nanocloud/models/users/users.go [hunk 5]
```diff
@@ -377,12 +400,12 @@ func GetUser(id string) (*User, error) {
...
 			&user.IsAdmin,
 			&user.Activated,
-			&user.SignupDate,
+			&timestamp,
 		)
 		if err != nil {
 			return nil, err
 		}
-		user.SignupDate *= 1000
+		user.SignupDate = int(1000 * timestamp)
```
Comment (Gentux) [inline nanocloud/models/users/users.go:410]: Same comment as above

File: nanocloud/models/users/users.go [hunk 6]
```diff
@@ -292,6 +294,26 @@ func UpdateUserPassword(id string, password string) error {
 	return nil
 }
 
+func UpdateUserRank(id string, rank bool) error {
```
Comment (Gentux) [inline nanocloud/models/users/users.go]: Rename this as "UpdateUserPrivilge"

File: nanocloud/routes/users/users.go [hunk 7]
```diff
@@ -127,47 +127,61 @@ func Disable(userId string) (int, error) {
...
-		err = users.UpdateUserPassword(user.GetID(), u.Password)
+	if !user.IsAdmin && (updatedUser.GetID() != user.GetID()) {
+		return apiErrors.Unauthorized.Detail("You can only update your account")
+	}
+
+	if updatedUser.IsAdmin != currentUser.IsAdmin {
+		if currentUser.Id == user.GetID() {
+			return apiErrors.Unauthorized.Detail("You cannot grant administration rights")
+		}
+		err = users.UpdateUserRank(updatedUser.GetID(), updatedUser.IsAdmin)
```
Comment (Gentux) [inline nanocloud/routes/users/users.go]: You forgot to rename this call to _UpdateUserPrivilege_
I suggest you to use 

```
git grep UpdateUserRank
```

**Pull request merged**